src/config/config-factory.ts

```diff
@@ -1,24 +1,64 @@
-import { PartnerConfig } from '.';
+import { ConfigSystem, PartnerConfig } from '.';
 import kenyaConfig from './chis-ke';
 import togoConfig from './chis-tg';
 import civConfig from './chis-civ';
 import maliConfig from './chis-ml';
+import path from 'path';
+import fs from 'fs';
 
-export const CONFIG_MAP: { [key: string]: PartnerConfig } = {
+export const uploadedConfigFilePath: string = path.join(getConfigUploadDirectory(), 'config.json');
+
+
+export const DEFAULT_CONFIG_MAP: { [key: string]: PartnerConfig } = {
   'CHIS-KE': kenyaConfig,
   'CHIS-TG': togoConfig,
   'CHIS-CIV': civConfig,
   'CHIS-ML': maliConfig
 };
 
-export default function getConfigByKey(key: string = 'CHIS-KE'): PartnerConfig {
+export default async function getConfigByKey(key: string = 'CHIS-KE'): Promise<PartnerConfig> {
+  if (fs.existsSync(uploadedConfigFilePath)) {
+    const uploadedConfig = await readConfig();
+    console.log(`Using uploaded configuration: ${uploadedConfigFilePath}`);
+    return uploadedConfig;
+  }
+
   const usingKey = key.toUpperCase();
   console.log(`Using configuration: ${key}`);
-  const result = CONFIG_MAP[usingKey];
+  const result = DEFAULT_CONFIG_MAP[usingKey];
   if (!result) {
-    const available = JSON.stringify(Object.keys(CONFIG_MAP));
+    const available = JSON.stringify(Object.keys(DEFAULT_CONFIG_MAP));
     throw Error(`Failed to start: Cannot find configuration "${usingKey}". Configurations available are ${available}`);
   }
 
   return result;
 }
+
+export function getConfigUploadDirectory (): string {
+  const configDir = path.join(__dirname, '..', 'config_uploads');
+  if (!fs.existsSync(configDir)) {
+    fs.mkdirSync(configDir);
+  }
+  return configDir;
+}
+
+export async function writeConfig(jsonConfigData: ConfigSystem): Promise<void> {
+  try {
+    const jsonString: string = JSON.stringify(jsonConfigData, null, 2);
+    await fs.promises.writeFile(uploadedConfigFilePath, jsonString);
+  } catch (error) {
+    throw new Error('writeConfig: Failed to write file');
+  }
+}
+
+export async function readConfig(): Promise<PartnerConfig> {
+  try {
+    const fileContent = await fs.promises.readFile(uploadedConfigFilePath, 'utf-8');
```

Comment:
I see alot of calls had to be converted to async so i have to ask, Is there a synchronous equivalent to this?

Reply:
yeah sure there is but its asynchronous mainly for performance.
since this is an api/server you'd want to support concurrency with multiple requests without blocking
also config is read almost every time you make a request to the server for a page

Reply:
If It's for performance, wouldn't caching the config be more performant than reading from disk for every request?

Reply:
maybe maybe not
but generally i wouldn't advice reading file synchronously in a server cos of its blocking nature
It more like reading db data synchronously, not advisable